node/actors/bft/src/leader/replica_commit.rs

```diff
@@ -96,14 +96,18 @@ impl StateMachine {
             return Err(Error::NotLeaderInView);
         }
 
+        // Get current incrementally-constructed QC to work on it
+        let commit_qc = self
+            .commit_qcs
+            .entry(message.view.number)
+            .or_default()
+            .entry(message.clone())
+            .or_insert_with(|| CommitQC::new(message.clone(), self.config.genesis()));
+
         // If we already have a message from the same validator and for the same view, we discard it.
-        if let Some(existing_message) = self
-            .commit_message_cache
-            .get(&message.view.number)
-            .and_then(|x| x.get(author))
-        {
+        if commit_qc.signers.0[self.config.genesis().validators.index(author).unwrap()] {
             return Err(Error::DuplicateMessage {
-                existing_message: existing_message.msg.clone(),
+                existing_message: commit_qc.message.clone(),
             });
```

Comment:
it is not an issue that a replica signs the same ReplicaCommit twice - CommitQC::add filters out the duplicates (see the implementation). The problem is that the same replica might sign DIFFERENT ReplicaCommit for the same view multiple times and THAT's what we should check here.

Reply:
Ok. Just to be sure, this was the same in the previous check with the `commit_message_cache`? 
I mean, it's not that new code behaves differently: Same `ReplicaCommit` signed twice by the same replica was raising `DuplicateError` in previous version as well, right?

Reply:
Actually, existing test is expecting an `Error::DuplicateMessage` when the `ReplicaCommit`s [match](https://github.com/matter-labs/era-consensus/blob/2f581685b5bdeb854b1d697966585ee7c517ff3c/node/actors/bft/src/leader/tests.rs#L555).
Is that test wrong, so?

Reply:
Ok @pompon0, @brunoffranca, need clarification for this:

**Previous version**: checks if same validator signed twice for same view **any** message        (will return error if signs _different_ or _same_ message).
 **Current version**: checks if same validator signed twice for same view **the same** message   (will return error _only_ if it signs _same_ message).
**Desired version?**: checks if same validator signed twice for same view **different** messages (will return error _only_ if it signs _different_ messages).

Do we agree we want **Desired version?**

Reply:
Either desired or previous version is okay.

Reply:
yes, desired and previous are okay. My point was to just reduce the amount of information we store (previous implementation stored messages both in aggregated and non-aggregated form, while it would be enough to remember aggregated form + for each validator what was the highest view of messages sent by them).